Documentation example for deploying to Google Cloud with Static Web Apps (#2740)
commercetools/merchant-center-application-kit: website/src/content/deployment-examples/google-cloud-storage.mdx

---
title: Deploying to Google cloud
---

This deployment example refers to [Google cloud Static Web Apps](https://codelabs.developers.google.com/codelabs/cloud-webapp-hosting-gcs#0)

# Prerequisites

Before you get started, you need to have:

> - A [google](https://accounts.google.com/SignUp) account.

Reviewer: ```suggestion
- A [Google](https://accounts.google.com/SignUp) account.
```

- A custom domain
- A [commercetools](https://docs.commercetools.com/getting-started) account and a Project.
- A Custom Application [configured in the Merchant Center](https://docs.commercetools.com/merchant-center/managing-custom-applications).

# Configuration

In your Custom Application config, provide the [Custom Application ID](/api-reference/application-config#envproductionapplicationid) you got when you configured the Custom Application in the Merchant Center.

```json title="custom-application-config.json"
{
  "env": {
    "production": {
      "applicationId": "ckvtahxl90097sys6har1e6n3",
      "url": "https://<custom_domain>"
    }
  }
}
```

## Using environment variables

In case you want to avoid hardcoding certain values, for example the Application ID, or the Project key, you can [use variable placeholders](/api-reference/application-config#using-variable-placeholders) in your Custom Application config.

```json title="custom-application-config.json" highlightLines="4"
{
  "env": {
    "production": {
      "applicationId": "${env:APPLICATION_ID}",
      "url": "https://<custom_domain>"
    }
  }
}
```

# Set up Google Cloud

> To host the production bundles of your Custom Application you need to create a [Cloud Storage bucket](https://console.cloud.google.com/storage) in Google cloud and [configure](https://cloud.google.com/storage/docs/hosting-static-website#create-bucket) the bucket.

Reviewer: ```suggestion
To host the production bundles of your Custom Application you need to create a [Cloud Storage bucket](https://console.cloud.google.com/storage) in Google Cloud and [configure](https://cloud.google.com/storage/docs/hosting-static-website#create-bucket) the bucket.
```

> To host the production bundles of your Custom Application you need to create a [Cloud Storage bucket](https://console.cloud.google.com/storage) in Google cloud and [configure](https://cloud.google.com/storage/docs/hosting-static-website#create-bucket) the bucket.

Reviewer: Is there anything in particular that needs to be configured in the bucket? If so, we should document that.

Make sure you are the owner and administrator of your domain.

<Info>

Merchant Center only allows accessing your Custom Application using a secure protocol (HTTPS) so your domain must support it with a valid [SSL certificate](https://support.google.com/domains/answer/7630973?hl=en&visit_id=637951191082633113-3999318342&rd=1).

> </Info>

Reviewer: I think we can include this into its own section about having a domain with a valid TLS certificate, as Google does not provide one automatically.

> # Connect Google cloud with GitHub Actions

Reviewer: ```suggestion
# Connect Google Cloud with GitHub
```

One of the ways to deploy to google cloud is to use [Cloud Build](https://cloud.google.com/build/docs/automating-builds/create-manage-triggers#connect_repo) deployment service. This service enables GitHub integration.

<Info>

Make sure to also install the [Google Cloud GitHub App](https://github.com/apps/google-cloud-build) to grant access to your repository.

> </Info>

Reviewer: I'm not sure we should recommend using Cloud Build, as this is usually for building docker images etc.

For static files, you just need to upload them to the bucket. Maybe we should recommend using a github action: https://github.com/google-github-actions/upload-cloud-storage

So the github integration could be about setting up the github action to upload the static files to the bucket.

Author: Thank you, all updated here: https://github.com/commercetools/merchant-center-application-kit/pull/2740/commits/8dd0ba6dfdccc43d45295f7b963abbef68ebf649

---
title: Deploying to Google cloud
---

This deployment example refers to [Google cloud Static Web Apps](https://codelabs.developers.google.com/codelabs/cloud-webapp-hosting-gcs#0)

# Prerequisites

Before you get started, you need to have:

- A [Google](https://accounts.google.com/SignUp) account.

> - A custom domain

Reviewer: ```suggestion
- A [valid domain](https://codelabs.developers.google.com/codelabs/cloud-webapp-hosting-gcs#1) that you own or manage.
```

# Configuration

In your Custom Application config, provide the [Custom Application ID](/api-reference/application-config#envproductionapplicationid) you got when you configured the Custom Application in the Merchant Center.

```json title="custom-application-config.json"
{
  "env": {
    "production": {
      "applicationId": "ckvtahxl90097sys6har1e6n3",
      "url": "https://<custom_domain>"
    }
  }
}
```

## Using environment variables

In case you want to avoid hardcoding certain values, for example the Application ID, or the Project key, you can [use variable placeholders](/api-reference/application-config#using-variable-placeholders) in your Custom Application config.

```json title="custom-application-config.json" highlightLines="4"
{
  "env": {
    "production": {
      "applicationId": "${env:APPLICATION_ID}",
      "url": "https://<custom_domain>"
    }
  }
}
```

# Set up Google Cloud

To host the production bundles of your Custom Application you need to create a [Cloud Storage bucket](https://console.cloud.google.com/storage) in Google Cloud and [configure](https://codelabs.developers.google.com/codelabs/cloud-webapp-hosting-gcs#3) the bucket.

> # Configuration

Reviewer: ```suggestion
## Configuring the bucket
```

> Depending on the access options you choose, if you want to configure permissions, select the created bucket, go to the permissions tab and click the `Add` link. For the `New principals` field, select `AllUsers` and for the `Role`, search for `Storage Object Viewer`. Click `Save` to make bucket publicly accessible.

Reviewer: ```suggestion
Depending on the access options you choose, if you want to configure permissions, select the created bucket, go to the permissions tab and click the `Add` link. For the `New principals` field, select `AllUsers` and for the `Role`, search for `Storage Object Viewer`. Click `Save` to make the bucket publicly accessible.
```

> # Have a valid domain
> 
> Make sure you are the owner and administrator of your domain.
> 
> Merchant Center only allows accessing your Custom Application using a secure protocol (HTTPS) so your domain must support it with a valid [SSL certificate](https://support.google.com/domains/answer/7630973?hl=en&visit_id=637951191082633113-3999318342&rd=1). Google does not provide one automatically

Reviewer: ```suggestion
# Domain verification

Creating buckets with domain names requires to verify the ownership of the domain.

<Info>

Google does not provide a domain by default when using Google Cloud Storage. If you don't have an existing domain, then there are many services through which you can register a new domain, such as [Google Domains](https://domains.google.com/).

</Info>

We recommend following the instructions for [Domain-named bucket verification](https://cloud.google.com/storage/docs/domain-name-verification) to configure things properly.
```

Author: Thank you. 
Feedback updated here: https://github.com/commercetools/merchant-center-application-kit/pull/2740/commits/7fd895792887039dc496fa96b8804ef3f03a97b1
I took the opportunity to emphasize some configurations on the load balancer to help link the SSL certificate to the domain.
Finally have the Custom app up and working as expected 😅

Reviewer: Nice!

# Connect Google cloud with GitHub Actions

> One of the ways to deploy to google cloud is to use [Github actions](https://github.com/google-github-actions/upload-cloud-storage) deployment service. This service enables GitHub integration.

Reviewer: ```suggestion
One of the ways to deploy to Google Cloud Storage (GCS) is to use the [Cloud Storage Uploader](https://github.com/google-github-actions/upload-cloud-storage) GitHub Action. Follow the instructions to integrate it with your repository CI.
```

<Info>

Make sure to also install the Google Cloud Storage uploader [GitHub App](https://github.com/marketplace/actions/cloud-storage-uploader) to grant access to your repository.

> </Info>

Reviewer: Hmm this is not a github app, it's a github action. Since you already mention it above, this can be removed I think.

## Configuring build settings

In the Google Cloud setup process you need to configure the following things:

* Override the *Build command* with: `yarn build`.

  <Info>

  In your `package.json` make sure to have the following scripts defined:

  ```json title="package.json"
  {
    "scripts": {
      "build": "mc-scripts build"
    }
  }
  ```

>   </Info>

Reviewer: I think this can be removed now

- A [google](https://accounts.google.com/SignUp) account.
- A custom domain
- A [commercetools](https://docs.commercetools.com/getting-started) account and a Project.
- A Custom Application [configured in the Merchant Center](https://docs.commercetools.com/merchant-center/managing-custom-applications).

# Configuration

In your Custom Application config, provide the [Custom Application ID](/api-reference/application-config#envproductionapplicationid) you got when you configured the Custom Application in the Merchant Center.

```json title="custom-application-config.json"
{
  "env": {
    "production": {
      "applicationId": "ckvtahxl90097sys6har1e6n3",
      "url": "https://<custom_domain>"
    }
  }
}
```

## Using environment variables

In case you want to avoid hardcoding certain values, for example the Application ID, or the Project key, you can [use variable placeholders](/api-reference/application-config#using-variable-placeholders) in your Custom Application config.

```json title="custom-application-config.json" highlightLines="4"
{
  "env": {
    "production": {
      "applicationId": "${env:APPLICATION_ID}",
      "url": "https://<custom_domain>"
    }
  }
}

> ```

Reviewer 2: I feel I'm missing a fragment where we configure those environment variables somewhere in Google's infrastructure (like the ones we have in other pages starting with `If your Custom Application config requires environment variables, make sure to provide them in...`). 

And if there is actually no option to do that in case of Cloud Storage, perhaps we should provide an example how to that using e.g. GitHub Actions or consider removing this fragment.

Reviewer 2: Thanks for the changes. It now occurred to me I was wrong, we're documenting a fully manual deployment process. User is supposed to build the app locally using local `.env` file and then manually upload it. As much as integration with GitHub Actions is a nice addition in general, my suggestion didn't make much sense. Sorry for that.

Author: Totally fine! Thanks for circling back on this 😁

---

> title: Deploying to Google cloud

Reviewer: ```suggestion
title: Deploying to Google Cloud with Static Web Apps
```

---

> This deployment example refers to [Google cloud Static Web Apps](https://codelabs.developers.google.com/codelabs/cloud-webapp-hosting-gcs#0)

Reviewer: ```suggestion
This deployment example refers to [Google Cloud Static Web Apps](https://codelabs.developers.google.com/codelabs/cloud-webapp-hosting-gcs#0)
```

# Prerequisites

Before you get started, you need to have:

- A [Google](https://accounts.google.com/SignUp) account.

> - A [valid domain](https://codelabs.developers.google.com/codelabs/cloud-webapp-hosting-gcs#1) that you own or manage.- A [commercetools](https://docs.commercetools.com/getting-started) account and a Project.

Reviewer: ```suggestion
- A [valid domain](https://codelabs.developers.google.com/codelabs/cloud-webapp-hosting-gcs#1) that you own or manage.
- A [commercetools](https://docs.commercetools.com/getting-started) account and a Project.
```

Author: Thank you. Updated here: https://github.com/commercetools/merchant-center-application-kit/pull/2740/commits/c197b5db5d2ae0c53e15e3b05abc4661e2b4cb21
